File: tools/journal.js
// The journal + snapshot persistence engine.
//
// `chronolog.chronolog` is a snapshot: a plain serialized document. Every
// committed edit appends one JSONL line to `chronolog.journal` in the same
// data directory. Loading is snapshot-plus-replay; compaction folds the
// journal back into the snapshot and truncates it.
//
// This module holds ZERO domain knowledge. An op names a map and a record id
// and either assigns a JSON value or deletes it — `document[map][id] = value`
// or `delete document[map][id]`, uniformly for all seven maps. `meta` and
// `foreign` are records keyed by their own top-level property name, which is
// exactly why no map needs special handling here. That uniformity is what
// makes server-side compaction possible without teaching the server what an
// event, a frame, or an ICS source is.

import { open, readFile, rename, unlink, writeFile } from "node:fs/promises";
import { join } from "node:path";
import { RECORD_MAPS, applyOps } from "../src/ops.js";

// What an op means is defined once, in src/ops.js, and imported by both sides.
// The client replays a conflicting window's ops with the same function the
// server replays the journal with; two implementations that drifted apart
// would corrupt documents quietly.
export { applyOps };
export const OP_MAPS = RECORD_MAPS;
const OP_MAP_SET = new Set(OP_MAPS);
export const DEFAULT_SNAPSHOT_PERIOD_MINUTES = 10;

export function validateEntryOps(ops) {
  if (!Array.isArray(ops)) throw new TypeError("entry.ops must be an array");
  for (const op of ops) {
    if (!op || typeof op !== "object") throw new TypeError("journal op must be an object");
    if (!OP_MAP_SET.has(op.map)) throw new RangeError(`unknown journal map "${op.map}"`);
    if (typeof op.id !== "string" || !op.id) throw new TypeError("journal op needs a record id");
    if (op.op === "put") {
      if (op.value === undefined) throw new TypeError(`put op for ${op.map}/${op.id} carries no value`);
    } else if (op.op !== "del") {
      throw new RangeError(`unknown journal op "${op.op}"`);
    }
  }
  return ops;
}

// Parses a journal buffer into entries. A process killed between `write` and
// `fsync` can leave a partial final line; that line is discarded with a
// warning and the healthy byte length is reported so the caller can truncate
// the file back to it. A damaged interior line is skipped with a warning too —
// losing one edit is bad, but refusing to open the document is worse.
export function parseJournal(text) {
  const entries = [];
  const warnings = [];
  let healthyBytes = 0;
  const lines = text.split("\n");
  for (let index = 0; index < lines.length; index += 1) {
    const line = lines[index];
    const isLast = index === lines.length - 1;
    if (!line) {
      if (!isLast) healthyBytes += 1;
      continue;
    }
    let entry = null;
    try {
      entry = JSON.parse(line);
    } catch {
      warnings.push(isLast
        ? `discarded a truncated final journal line (${line.length} bytes); the last append did not complete`
        : `skipped an unreadable journal line at line ${index + 1}`);
      if (!isLast) healthyBytes += Buffer.byteLength(line) + 1;
      continue;
    }
    if (!entry || typeof entry !== "object" || !Number.isInteger(entry.seq) || !Array.isArray(entry.ops)) {
      warnings.push(`skipped a malformed journal entry at line ${index + 1}`);
      if (!isLast) healthyBytes += Buffer.byteLength(line) + 1;
      continue;
    }
    entries.push(entry);
    healthyBytes += Buffer.byteLength(line) + (isLast ? 0 : 1);
  }
  return { entries, warnings, healthyBytes };
}

export function serializeEntry(entry) {
  return JSON.stringify(entry) + "\n";
}

export class JournalConflict extends Error {
  constructor(currentSeq, missed, truncated) {
    super(`journal is at seq ${currentSeq}`);
    this.name = "JournalConflict";
    this.currentSeq = currentSeq;
    this.missed = missed;
    this.truncated = truncated;
  }
}

async function atomicWrite(dataRoot, file, body) {
  const temporary = join(
    dataRoot,
    `.chronolog-save-${process.pid}-${Date.now()}-${Math.random().toString(16).slice(2)}.tmp`
  );
  const descriptor = await open(temporary, "wx", 0o600);
  try {
    await descriptor.writeFile(body);
    await descriptor.sync();
  } finally {
    await descriptor.close();
  }
  try {
    await rename(temporary, file);
  } finally {
    await unlink(temporary).catch(() => {});
  }
}

export function createJournalStore({ dataRoot, warn = () => {} }) {
  const snapshotFile = join(dataRoot, "chronolog.chronolog");
  const journalFile = join(dataRoot, "chronolog.journal");
  const stateFile = join(dataRoot, ".chronolog-journal-state.json");
  const settingsFile = join(dataRoot, ".chronolog-settings.json");

  let document = null;
  let currentSeq = 0;
  // Entries still in the journal file, kept so a lagging client can rebase.
  // Compaction clears this; `journalBase` then records the seq below which
  // history is no longer reconstructible and a full reload is required.
  let entries = [];
  let journalBase = 0;
  let serialized = null;
  let writes = Promise.resolve();
  let periodic = null;

  function invalidate() {
    serialized = null;
  }

  async function readSidecarSeq() {
    try {
      const parsed = JSON.parse(await readFile(stateFile, "utf8"));
      return {
        seq: Number.isInteger(parsed?.seq) ? parsed.seq : 0,
        base: Number.isInteger(parsed?.base) ? parsed.base : 0
      };
    } catch (error) {
      if (error?.code !== "ENOENT") warn(`journal state unreadable: ${error.message}`);
      return { seq: 0, base: 0 };
    }
  }

  async function load() {
    let snapshotText = null;
    try {
      snapshotText = await readFile(snapshotFile, "utf8");
    } catch (error) {
      if (error?.code !== "ENOENT") throw error;
    }
    let journalText = "";
    let journalPresent = false;
    try {
      journalText = await readFile(journalFile, "utf8");
      journalPresent = true;
    } catch (error) {
      if (error?.code !== "ENOENT") throw error;
    }
    if (snapshotText === null && !journalPresent) {
      document = null;
      currentSeq = 0;
      entries = [];
      journalBase = 0;
      invalidate();
      return { present: false, replayed: 0, warnings: [] };
    }

    const sidecar = await readSidecarSeq();
    const parsedJournal = parseJournal(journalText);
    for (const message of parsedJournal.warnings) warn(message);
    if (journalPresent && parsedJournal.healthyBytes !== Buffer.byteLength(journalText)) {
      // Heal the file so the next append does not build on a partial line.
      const handle = await open(journalFile, "r+");
      try {
        await handle.truncate(parsedJournal.healthyBytes);
        await handle.sync();
      } finally {
        await handle.close();
      }
    }

    if (snapshotText === null) {
      warn("no snapshot found; replaying the journal onto an empty document");
      document = {};
    } else {
      document = JSON.parse(snapshotText);
    }

    entries = parsedJournal.entries;
    let replayed = 0;
    for (const entry of entries) {
      applyOps(document, entry.ops);
      replayed += 1;
    }
    const lastJournalSeq = entries.length ? entries[entries.length - 1].seq : 0;
    // Both sources are authoritative in different crash windows: the sidecar
    // survives a truncated journal, the journal survives a lost sidecar.
    currentSeq = Math.max(sidecar.seq, lastJournalSeq);
    journalBase = entries.length ? entries[0].seq - 1 : Math.max(sidecar.base, currentSeq);
    invalidate();
    return { present: true, replayed, warnings: parsedJournal.warnings };
  }

  async function writeState() {
    await atomicWrite(dataRoot, stateFile, JSON.stringify({ seq: currentSeq, base: journalBase }) + "\n");
  }

  // Appends are serialized through one promise chain so seq assignment and
  // byte order can never interleave, and each append is fsynced before the
  // caller is told it committed.
  function append(baseSeq, incoming) {
    const run = writes.then(async () => {
      if (baseSeq !== currentSeq) {
        const missed = entries.filter((entry) => entry.seq > baseSeq);
        // History below journalBase was folded into the snapshot, so it can no
        // longer be replayed as ops; the client must reload instead.
        throw new JournalConflict(currentSeq, missed, baseSeq < journalBase);
      }
      for (const entry of incoming) validateEntryOps(entry?.ops);
      const stamped = [];
      let body = "";
      const timestamp = new Date().toISOString();
      let seq = currentSeq;
      for (const entry of incoming) {
        seq += 1;
        const line = { seq, ts: timestamp, label: String(entry.label ?? ""), ops: entry.ops };
        stamped.push(line);
        body += serializeEntry(line);
      }
      if (stamped.length) {
        const handle = await open(journalFile, "a", 0o600);
        try {
          await handle.writeFile(body);
          await handle.sync();
        } finally {
          await handle.close();
        }
        for (const entry of stamped) {
          applyOps(document, entry.ops);
          entries.push(entry);
        }
        currentSeq = seq;
        invalidate();
        await writeState();
      }
      return { seq: currentSeq };
    });
    writes = run.catch(() => {});
    return run;
  }

  function since(seq) {
    return {
      currentSeq,
      truncated: seq < journalBase,
      entries: entries.filter((entry) => entry.seq > seq)
    };
  }

  function body() {
    if (document === null) return null;
    // Rebuilt only when a GET actually arrives. Appends merely invalidate, so
    // a burst of edits costs no serialization at all.
    serialized ||= Buffer.from(JSON.stringify(document));
    return serialized;
  }

  // Compaction is the whole point of domain-free ops: fold the journal into
  // the snapshot, then truncate. Order matters — the snapshot must be durable
  // before the journal is dropped, and because ops are idempotent a crash
  // between the two simply replays entries that are already folded in.
  function compact({ force = false } = {}) {
    const run = writes.then(async () => {
      if (document === null) return { compacted: false, folded: 0 };
      if (!entries.length && !force) return { compacted: false, folded: 0 };
      const folded = entries.length;
      await atomicWrite(dataRoot, snapshotFile, JSON.stringify(document) + "\n");
      journalBase = currentSeq;
      await writeState();
      const handle = await open(journalFile, "w", 0o600);
      try {
        await handle.sync();
      } finally {
        await handle.close();
      }
      entries = [];
      return { compacted: true, folded };
    });
    writes = run.catch(() => {});
    return run;
  }

  // Whole-document replacement: the owner opening a different file. There is
  // no CAS here — it is an explicit, deliberate act, not an autosave. The
  // journal is dropped because its ops describe a document that no longer
  // exists, and seq advances rather than resetting so any other window's
  // baseSeq goes stale and it reloads instead of appending onto the wrong
  // document.
  function replaceSnapshot(nextDocument) {
    const run = writes.then(async () => {
      document = nextDocument;
      currentSeq += 1;
      journalBase = currentSeq;
      await atomicWrite(dataRoot, snapshotFile, JSON.stringify(document) + "\n");
      await writeState();
      const handle = await open(journalFile, "w", 0o600);
      try {
        await handle.sync();
      } finally {
        await handle.close();
      }
      entries = [];
      invalidate();
      return { seq: currentSeq };
    });
    writes = run.catch(() => {});
    return run;
  }

  async function readSettings() {
    try {
      const parsed = JSON.parse(await readFile(settingsFile, "utf8"));
      const minutes = Number(parsed?.snapshotPeriodMinutes);
      return {
        snapshotPeriodMinutes: Number.isFinite(minutes) && minutes > 0
          ? minutes
          : DEFAULT_SNAPSHOT_PERIOD_MINUTES
      };
    } catch (error) {
      if (error?.code !== "ENOENT") warn(`settings unreadable: ${error.message}`);
      return { snapshotPeriodMinutes: DEFAULT_SNAPSHOT_PERIOD_MINUTES };
    }
  }

  async function writeSettings(next) {
    const minutes = Number(next?.snapshotPeriodMinutes);
    if (!Number.isFinite(minutes) || minutes <= 0) throw new RangeError("snapshotPeriodMinutes must be a positive number");
    const settings = { snapshotPeriodMinutes: minutes };
    await writeFile(settingsFile, JSON.stringify(settings) + "\n", { mode: 0o600 });
    await schedule();
    return settings;
  }

  async function schedule() {
    if (periodic) clearInterval(periodic);
    const { snapshotPeriodMinutes } = await readSettings();
    periodic = setInterval(() => {
      compact().catch((error) => warn(`periodic compaction failed: ${error.message}`));
    }, snapshotPeriodMinutes * 60 * 1000);
    periodic.unref?.();
    return snapshotPeriodMinutes;
  }

  function stop() {
    if (periodic) clearInterval(periodic);
    periodic = null;
  }

  return {
    load,
    append,
    since,
    compact,
    replaceSnapshot,
    body,
    readSettings,
    writeSettings,
    schedule,
    stop,
    get document() { return document; },
    get seq() { return currentSeq; },
    get journalBase() { return journalBase; },
    get entryCount() { return entries.length; },
    get present() { return document !== null; }
  };
}

File: app/lib/store/journal.dart
// The on-disk truth: a plaintext snapshot, and an append-only journal beside it.
//
// `chronolog.chronolog` is the serialized document -- compact JSON with a
// trailing newline, so the plaintext document diffs. Every committed edit
// appends one JSONL line to `chronolog.journal`. Loading is snapshot plus
// replay. Compaction folds the journal back into the snapshot and empties it.
//
// SAVES ARE UPDATES, NOT OVERWRITES. This file is that ruling in code: ordinary
// editing never rewrites the document, it appends the ops that changed it. The
// whole document is written on exactly three occasions -- establishing a file
// that does not exist yet, compaction, and the owner deliberately replacing the
// document, which is also how a MOVE to a chosen save location writes its first
// snapshot -- and none of them is autosave.
//
// ZERO DOMAIN KNOWLEDGE. An op names a map and a record id and either assigns
// the whole record or deletes it, uniformly for all seven maps. That uniformity
// is what lets this fold a journal into a snapshot with no idea what an event, a
// frame or an ICS source is.
//
// SINGLE WRITER. One master isolate owns the document and this journal, so there
// is no sequence-number CAS, no conflict type, and no rebase loop here. Those
// existed to arbitrate between browser windows over HTTP, and that whole surface
// is ruled out of pass one.

import 'dart:convert';

import '../core/document.dart';
import '../core/ops.dart';
import '../core/records.dart';
import '../core/validate.dart';
import 'data_dir.dart';
import 'seams.dart';

const String snapshotFileName = 'chronolog.chronolog';
const String journalFileName = 'chronolog.journal';
const String journalStateFileName = '.chronolog-journal-state.json';

const int _newline = 0x0a;

/// The plaintext document writer. Compact, one trailing newline, and the record
/// order records.dart declares -- so two saves of the same document are the same
/// bytes and a diff shows only what changed.
String serializeDocument(Document document) => '${jsonEncode(document.toJson())}\n';

/// One committed edit as the journal stores it: its sequence number, when it was
/// written, its own label, and the ops that are the edit.
class JournalEntry {
  const JournalEntry({required this.seq, required this.at, required this.label, required this.ops});

  final int seq;
  final String at;
  final String label;
  final List<Op> ops;

  Json toJson() => {
    'seq': seq,
    'ts': at,
    'label': label,
    'ops': [for (final op in ops) op.toJson()],
  };

  /// Null when the line is not an entry. A line missing its seq or its ops is
  /// not a partially good entry, it is not an entry.
  static JournalEntry? tryFrom(Object? raw) {
    if (raw is! Map) return null;
    final seq = raw['seq'];
    final ops = raw['ops'];
    if (seq is! int || ops is! List) return null;
    return JournalEntry(
      seq: seq,
      at: '${raw['ts'] ?? ''}',
      label: '${raw['label'] ?? ''}',
      ops: [
        for (final op in ops)
          if (op is Map) Op.fromJson(Json.from(op)),
      ],
    );
  }
}

/// What a load found.
///
/// TWO KINDS OF NEWS, kept apart because they are answered differently.
/// [reports] is what the store had to do to the FILES -- a torn tail dropped, a
/// missing snapshot replayed onto an empty document -- and an empty list means
/// the files were whole. [findings] is what the validator says about the
/// DOCUMENT, in its own vocabulary; the owner reads those and decides. Neither
/// one ever refuses the load.
typedef JournalLoad = ({
  Document document,
  bool present,
  int replayed,
  int seq,
  List<String> reports,
  List<String> findings,
});

/// A journal buffer read as entries.
///
/// [healthyBytes] is the offset just past the last entry that parsed, so the
/// caller can cut the file back to it. A process killed mid-append leaves a
/// partial final line; that line is DROPPED with a report rather than taken as a
/// reason to refuse the document. An unreadable interior line is skipped with a
/// report and left in place -- losing one edit is bad, rewriting a journal to
/// tidy it is worse.
({List<JournalEntry> entries, List<String> reports, int healthyBytes}) parseJournal(
  List<int> bytes,
) {
  final entries = <JournalEntry>[];
  final reports = <String>[];
  var healthy = 0;
  var start = 0;
  var line = 0;
  while (start < bytes.length) {
    final found = bytes.indexOf(_newline, start);
    final torn = found < 0;
    final end = torn ? bytes.length : found;
    line += 1;
    if (end > start) {
      final entry = JournalEntry.tryFrom(_decode(bytes.sublist(start, end)));
      if (entry == null) {
        reports.add(
          torn
              ? 'Dropped a truncated final journal line (${end - start} bytes); '
                    'the last append did not complete.'
              : 'Skipped an unreadable journal line at line $line.',
        );
      } else {
        entries.add(entry);
        healthy = torn ? end : end + 1;
      }
    }
    start = end + 1;
  }
  return (entries: entries, reports: reports, healthyBytes: healthy);
}

/// A journal line as JSON, or null. Malformed UTF-8 is allowed through the
/// decoder because a torn tail can cut a multi-byte character in half, and a
/// decoder that threw there would turn a healable file into a dead one.
Object? _decode(List<int> bytes) {
  try {
    return jsonDecode(utf8.decode(bytes, allowMalformed: true));
  } on FormatException {
    return null;
  }
}

/// The snapshot-and-journal pair for one data directory.
///
/// It owns the FILES and the sequence number; it does not own the document. The
/// document is handed in when it must be written and handed back when it is
/// read, which is what keeps the live document in exactly one place -- the
/// master isolate's store above this.
class JournalStore {
  /// [continuing] is where this pair's numbering starts. It matters in exactly
  /// one case: a document MOVED to a new directory carries its sequence
  /// forward, so the history reads as one lineage rather than restarting at one.
  JournalStore({
    required this.dataRoot,
    this.files = const IoStoreFiles(),
    this.scheduler = const WallScheduler(),
    int continuing = 0,
  }) : _seq = continuing,
       snapshotFile = storePath(dataRoot, snapshotFileName),
       journalFile = storePath(dataRoot, journalFileName),
       _stateFile = storePath(dataRoot, journalStateFileName);

  final String dataRoot;
  final StoreFiles files;
  final Scheduler scheduler;
  final String snapshotFile;
  final String journalFile;
  final String _stateFile;

  int _seq;
  int _entryCount = 0;

  /// The sequence number of the last entry written. Monotone across a
  /// compaction, which is why the state sidecar exists at all.
  int get seq => _seq;

  /// Entries still in the journal file, folded into the snapshot by a compaction.
  int get entryCount => _entryCount;

  /// Does a chronolog already live here? Asked before a MOVE: a location
  /// another document occupies is not somewhere this one may be written, and
  /// with no confirmation dialogs anywhere the honest answer is to refuse and
  /// say so rather than to overwrite and hope.
  Future<bool> occupied() async =>
      await files.read(snapshotFile) != null ||
      (await files.read(journalFile) ?? const <int>[]).isNotEmpty;

  /// Snapshot plus replay.
  ///
  /// When neither file exists the document is ESTABLISHED -- [establish] mints
  /// it, and its snapshot is written so the journal has a base to append onto.
  /// Nothing is invented: the empty document is whatever the core's factory says
  /// an empty document is.
  Future<JournalLoad> load({Document Function()? establish}) async {
    await files.ensureDirectory(dataRoot);
    final snapshotBytes = await files.read(snapshotFile);
    final journalBytes = await files.read(journalFile) ?? const <int>[];
    final reports = <String>[];
    final parsed = parseJournal(journalBytes);
    reports.addAll(parsed.reports);
    await _healTail(journalBytes, parsed.healthyBytes);
    if (snapshotBytes == null && journalBytes.isEmpty) {
      final document = (establish ?? createEmptyWorkspaceDocument)();
      _seq = 0;
      _entryCount = 0;
      await _atomicWrite(snapshotFile, serializeDocument(document));
      await _writeState();
      return (
        document: document,
        present: false,
        replayed: 0,
        seq: 0,
        reports: reports,
        findings: validateDocument(document).errors,
      );
    }
    var document = const Document();
    if (snapshotBytes == null) {
      reports.add('No snapshot was found; the journal replayed onto an empty document.');
    } else {
      document = Document.fromJson(jsonDecode(_decodeText(snapshotBytes)) as Json);
    }
    for (final entry in parsed.entries) {
      document = applyOps(document, entry.ops);
    }
    final last = parsed.entries.isEmpty ? 0 : parsed.entries.last.seq;
    // Both sources are authoritative in different crash windows: the sidecar
    // survives a truncated journal, the journal survives a lost sidecar.
    _seq = _larger(await _readState(), last);
    _entryCount = parsed.entries.length;
    // The validator REPORTS. Migrations are dead by ruling and nothing here
    // rewrites the file to suit this build; what a load owes the owner is the
    // findings, in the law's own words, not a document that declines to open.
    return (
      document: document,
      present: true,
      replayed: parsed.entries.length,
      seq: _seq,
      reports: reports,
      findings: validateDocument(document).errors,
    );
  }

  /// Append committed edits, each keeping its own label, and flush before
  /// answering. Sequence numbers are assigned here and only here, one per entry,
  /// uniformly -- so the file's order and its numbering can never disagree.
  Future<int> append(List<OpEntry> entries) async {
    if (entries.isEmpty) return _seq;
    // The gate is the DEFINITION, not a restatement of it: an op this build
    // cannot replay must never reach the file, and applyOps is what "can
    // replay" means. An unknown verb, an unknown map or a record id that is
    // not one is refused here rather than persisted as a poison line.
    applyOps(const Document(), [for (final entry in entries) ...entry.ops]);
    final at = scheduler.now().toUtc().toIso8601String();
    final body = StringBuffer();
    var seq = _seq;
    for (final entry in entries) {
      seq += 1;
      body.writeln(
        jsonEncode(JournalEntry(seq: seq, at: at, label: entry.label, ops: entry.ops).toJson()),
      );
    }
    await files.appendSynced(journalFile, utf8.encode(body.toString()));
    _seq = seq;
    _entryCount += entries.length;
    await _writeState();
    return _seq;
  }

  /// Fold the journal into a new snapshot and empty it. Returns how many entries
  /// were folded.
  ///
  /// THE ORDER IS THE SAFETY. The snapshot is durable before the journal is
  /// dropped, so a crash in between leaves entries in the journal that are
  /// already folded into the snapshot -- and replaying them is harmless BECAUSE
  /// OPS ARE IDEMPOTENT. That is why a put carries the whole record rather than
  /// a field-level change: the same put applied twice lands on the same state,
  /// which is what makes a crash mid-compaction survivable at all.
  Future<int> compact(Document document, {bool force = false}) async {
    if (_entryCount == 0 && !force) return 0;
    final folded = _entryCount;
    await _snapshotAndReset(document);
    return folded;
  }

  /// Whole-document replacement: the owner opening a different document into
  /// this data directory. Never part of autosave. The journal is dropped because
  /// its ops describe a document that no longer exists, and the sequence number
  /// advances rather than resetting so the discontinuity is visible in the file.
  Future<void> replaceSnapshot(Document document) async {
    _seq += 1;
    await _snapshotAndReset(document);
  }

  Future<void> _snapshotAndReset(Document document) async {
    await _atomicWrite(snapshotFile, serializeDocument(document));
    await _writeState();
    await files.truncate(journalFile, 0);
    _entryCount = 0;
  }

  /// Write-new, flush, rename. The replacement is one filesystem step, so a
  /// reader either sees the whole old snapshot or the whole new one. The
  /// temporary is deleted first because a crash can leave one behind, and
  /// building on a leftover is the one thing an exclusive create refuses.
  Future<void> _atomicWrite(String path, String text) async {
    final temporary = '$path.tmp';
    await files.delete(temporary);
    await files.writeNew(temporary, utf8.encode(text));
    await files.rename(temporary, path);
  }

  /// Cut a torn tail off so the next append cannot land after a partial line,
  /// and guarantee the file ends on a newline so it cannot land inside a good
  /// one either.
  Future<void> _healTail(List<int> bytes, int healthy) async {
    if (bytes.isEmpty) return;
    if (healthy != bytes.length) await files.truncate(journalFile, healthy);
    if (healthy > 0 && bytes[healthy - 1] != _newline) {
      await files.appendSynced(journalFile, const [_newline]);
    }
  }

  Future<int> _readState() async {
    final bytes = await files.read(_stateFile);
    if (bytes == null) return 0;
    final parsed = _decode(bytes);
    final seq = parsed is Map ? parsed['seq'] : null;
    return seq is int ? seq : 0;
  }

  Future<void> _writeState() => _atomicWrite(_stateFile, '${jsonEncode({'seq': _seq})}\n');

  String _decodeText(List<int> bytes) => utf8.decode(bytes, allowMalformed: true);

  static int _larger(int a, int b) => a > b ? a : b;
}
